File: backend/app/modules/intermediaries_filing/progress.py
"""Stage-aware section completion for Intermediaries & Filing."""

from __future__ import annotations

from typing import Any

from app.modules.intermediaries_filing.constants import IF_CONFIRMATION_FIELDS, SECTION_IDS
from app.modules.intermediaries_filing.rules import is_stage_at_least


def _filled(value: Any) -> bool:
    if value is None:
        return False
    if isinstance(value, str):
        return value.strip() != ""
    if isinstance(value, bool):
        return value
    if isinstance(value, list):
        return len(value) > 0
    return True


def _status_from(answered: int, total: int, extra_complete: bool = True) -> str:
    if answered == 0:
        return "not_started"
    if answered < total or not extra_complete:
        return "in_progress"
    return "complete"


def _current_filing_stage(payload: dict[str, Any]) -> str:
    config = payload.get("issueConfigurationAndFilingSnapshot") or {}
    return str((config.get("filingSnapshot") or {}).get("filingStage") or "")


def evaluate_issue_team_status(payload: dict[str, Any]) -> str:
    section = payload.get("issueTeamAndIntermediaryMaster") or {}
    snapshot = section.get("issueTeamSnapshot") or {}
    intermediaries = section.get("intermediaries") or []
    core = [
        _filled(snapshot.get("teamAsOfDate")),
        _filled(snapshot.get("leadManagerAppointed")),
        len(intermediaries) > 0,
    ]
    answered = sum(1 for item in core if item)
    intermediaries_complete = all(
        _filled(item.get("legalName")) or _filled(item.get("displayName"))
        for item in intermediaries
        if isinstance(item, dict)
    )
    return _status_from(answered, len(core), intermediaries_complete)


def evaluate_issue_configuration_status(payload: dict[str, Any]) -> str:
    section = payload.get("issueConfigurationAndFilingSnapshot") or {}
    filing_snapshot = section.get("filingSnapshot") or {}
    reconciliation = section.get("filingSnapshotReconciliation") or {}
    core = [
        _filled(filing_snapshot.get("snapshotDate")),
        _filled(filing_snapshot.get("filingStage")),
        _filled(filing_snapshot.get("currentOfferDocumentForm")),
        _filled(reconciliation.get("filingConfirmationStatus")),
    ]
    answered = sum(1 for item in core if item)
    return _status_from(answered, len(core))


def evaluate_filing_tracker_status(payload: dict[str, Any]) -> str:
    section = payload.get("filingAndRegulatoryMilestoneTracker") or {}
    has_data = any(
        [
            section.get("filings"),
            _filled((section.get("exchangeDraftFiling") or {}).get("exchange")),
            _filled((section.get("inPrincipleApproval") or {}).get("applied")),
            _filled((section.get("sebiSmeFiling") or {}).get("filingApplicability")),
            _filled((section.get("rocFiling") or {}).get("filingDate")),
        ]
    )
    if not has_data:
        return "not_started"
    filings_complete = all(
        _filled(filing.get("documentType")) or _filled(filing.get("filingDate"))
        for filing in (section.get("filings") or [])
        if isinstance(filing, dict)
    )
    return "complete" if filings_complete else "in_progress"


def evaluate_due_diligence_status(payload: dict[str, Any]) -> str:
    section = payload.get("dueDiligenceCertificatesConsentsAndSignoffs") or {}
    has_data = any(
        [
            section.get("dueDiligenceAreas"),
            section.get("certificates"),
            section.get("consents"),
            section.get("chapterSignoffs"),
        ]
    )
    if not has_data:
        return "not_started"
    certificates_complete = all(
        _filled(certificate.get("certificateType")) or _filled(certificate.get("provider"))
        for certificate in (section.get("certificates") or [])
        if isinstance(certificate, dict)
    )
    return "complete" if certificates_complete else "in_progress"


def evaluate_infrastructure_status(payload: dict[str, Any]) -> str:
    section = payload.get("depositoriesBankingAsbaUpiAndIssueInfrastructure") or {}
    core = [
        _filled((section.get("depositoryReadiness") or {}).get("isinStatus")),
        _filled((section.get("sponsorBankUpiReadiness") or {}).get("sponsorBankAppointed")),
        _filled((section.get("asbaConfiguration") or {}).get("asbaApplicable")),
        len(section.get("issueBankRoles") or []) > 0,
    ]
    answered = sum(1 for item in core if item)
    if answered == 0:
        return "not_started"
    return _status_from(answered, len(core))


def evaluate_underwriting_status(payload: dict[str, Any]) -> str:
    section = payload.get("underwritingMarketMakingAndDistributionArrangements") or {}
    summary = section.get("underwritingSummary") or {}
    has_data = any(
        [
            _filled(summary.get("issueShares")),
            _filled(summary.get("totalUnderwritingCommitment")),
            section.get("underwritingCommitments"),
            _filled((section.get("marketMakerConfiguration") or {}).get("marketMakerIntermediaryId")),
        ]
    )
    if not has_data:
        return "not_started"
    commitments_complete = all(
        _filled(commitment.get("intermediaryId")) or _filled(commitment.get("sharesUnderwritten"))
        for commitment in (section.get("underwritingCommitments") or [])
        if isinstance(commitment, dict)
    )
    return "complete" if commitments_complete else "in_progress"


def evaluate_issue_programme_status(payload: dict[str, Any]) -> str:
    stage = _current_filing_stage(payload)
    section = payload.get("issueProgrammeAllotmentListingAndPostIssueExecution") or {}
    calendar = section.get("issueCalendar") or {}

    if not is_stage_at_least(stage, "issue_open"):
        calendar_started = _filled(calendar.get("issueOpeningDate")) or _filled(calendar.get("issueClosingDate"))
        return "in_progress" if calendar_started else "not_started"

    if not is_stage_at_least(stage, "issue_closed"):
        opening = section.get("issueOpeningReadiness") or {}
        answered = sum(
            1
            for key in ("rhpProspectusRocFilingReady", "pricingFinalized", "registrarReady")
            if opening.get(key) != ""
        )
        if answered == 0:
            return "not_started"
        return "complete" if answered >= 3 else "in_progress"

    if not is_stage_at_least(stage, "allotment"):
        return "in_progress" if section.get("subscriptionRows") else "not_yet_due"

    if not is_stage_at_least(stage, "listing_application"):
        basis = section.get("basisOfAllotment") or {}
        answered = sum(
            1
            for key in ("basisPrepared", "allotmentFinalized", "exchangeApprovalReceived")
            if basis.get(key) != ""
        )
        if answered == 0:
            return "not_yet_due"
        return "complete" if answered >= 3 else "in_progress"

    listing = section.get("listing") or {}
    listing_started = _filled(listing.get("finalListingApplicationSubmitted")) or _filled(
        listing.get("listingDate")
    )
    return "in_progress" if listing_started else "not_yet_due"


def evaluate_final_document_status(payload: dict[str, Any]) -> str:
    section = payload.get("finalOfferDocumentAdvertisementsMaterialDocumentsAndFilingReadiness") or {}
    confirmations = section.get("finalConfirmations") or {}
    answered = sum(1 for key, _ in IF_CONFIRMATION_FIELDS if confirmations.get(key) != "")
    if answered == 0 and not section.get("offerDocumentVersions"):
        return "not_started"
    if answered < len(IF_CONFIRMATION_FIELDS):
        return "in_progress"
    return "complete"


SECTION_EVALUATORS = {
    "issue-team-and-intermediary-master": evaluate_issue_team_status,
    "issue-configuration-and-filing-snapshot": evaluate_issue_configuration_status,
    "filing-and-regulatory-milestone-tracker": evaluate_filing_tracker_status,
    "due-diligence-certificates-consents-and-signoffs": evaluate_due_diligence_status,
    "depositories-banking-asba-upi-and-issue-infrastructure": evaluate_infrastructure_status,
    "underwriting-market-making-and-distribution-arrangements": evaluate_underwriting_status,
    "issue-programme-allotment-listing-and-post-issue-execution": evaluate_issue_programme_status,
    "final-offer-document-advertisements-material-documents-and-filing-readiness": (
        evaluate_final_document_status
    ),
}


def _derive_overall_status(sections: dict[str, str]) -> str:
    actionable = [
        status for status in sections.values() if status not in {"not_yet_due", "not_applicable"}
    ]
    if not actionable:
        return "not_started"
    complete_count = sum(1 for status in actionable if status == "complete")
    if complete_count == 0:
        return "not_started"
    if complete_count == len(actionable):
        return "complete"
    return "in_progress"


def calculate_intermediaries_filing_progress(payload: dict[str, Any]) -> dict[str, Any]:
    sections = {section_id: SECTION_EVALUATORS[section_id](payload) for section_id in SECTION_IDS}
    sections_complete = sum(1 for status in sections.values() if status == "complete")
    return {
        "sections": sections,
        "sectionsComplete": sections_complete,
        "totalSections": len(sections),
        "overallStatus": _derive_overall_status(sections),
        "currentFilingStage": _current_filing_stage(payload),
    }

File: frontend/lib/intermediaries-filing/progress.ts
/**
 * Stage-aware section completion for Intermediaries & Filing.
 */

import { IF_CONFIRMATION_FIELDS } from '@/lib/intermediaries-filing/options';
import { isStageAtLeast } from '@/lib/intermediaries-filing/rules';
import type {
  IfProgress,
  IntermediariesFilingSectionId,
  SectionStatus,
} from '@/lib/intermediaries-filing/types';
import type { IntermediariesFilingPayload, FilingStage } from '@/lib/schemas/intermediaries-filing';

function filled(value: unknown): boolean {
  if (value === null || value === undefined) return false;
  if (typeof value === 'string') return value.trim().length > 0;
  if (typeof value === 'boolean') return value;
  if (Array.isArray(value)) return value.length > 0;
  return true;
}

function statusFrom(answered: number, total: number, extraComplete = true): SectionStatus {
  if (answered === 0) return 'not_started';
  if (answered < total || !extraComplete) return 'in_progress';
  return 'complete';
}

function getCurrentFilingStage(payload: IntermediariesFilingPayload): FilingStage | '' {
  return payload.issueConfigurationAndFilingSnapshot.filingSnapshot.filingStage;
}

export function evaluateIssueTeamStatus(payload: IntermediariesFilingPayload): SectionStatus {
  const section = payload.issueTeamAndIntermediaryMaster;
  const snapshot = section.issueTeamSnapshot;
  const core = [
    filled(snapshot.teamAsOfDate),
    filled(snapshot.leadManagerAppointed),
    section.intermediaries.length > 0,
  ];
  const answered = core.filter(Boolean).length;
  const intermediariesComplete = section.intermediaries.every(
    (intermediary) => filled(intermediary.legalName) || filled(intermediary.displayName),
  );
  return statusFrom(answered, core.length, intermediariesComplete);
}

export function evaluateIssueConfigurationStatus(
  payload: IntermediariesFilingPayload,
): SectionStatus {
  const section = payload.issueConfigurationAndFilingSnapshot;
  const core = [
    filled(section.filingSnapshot.snapshotDate),
    filled(section.filingSnapshot.filingStage),
    filled(section.filingSnapshot.currentOfferDocumentForm),
    filled(section.filingSnapshotReconciliation.filingConfirmationStatus),
  ];
  const answered = core.filter(Boolean).length;
  return statusFrom(answered, core.length);
}

export function evaluateFilingTrackerStatus(payload: IntermediariesFilingPayload): SectionStatus {
  const section = payload.filingAndRegulatoryMilestoneTracker;
  const hasData =
    section.filings.length > 0 ||
    filled(section.exchangeDraftFiling.exchange) ||
    filled(section.inPrincipleApproval.applied) ||
    filled(section.sebiSmeFiling.filingApplicability) ||
    filled(section.rocFiling.filingDate);
  if (!hasData) return 'not_started';

  const filingsComplete = section.filings.every(
    (filing) => filled(filing.documentType) || filled(filing.filingDate),
  );
  return filingsComplete ? 'complete' : 'in_progress';
}

export function evaluateDueDiligenceStatus(payload: IntermediariesFilingPayload): SectionStatus {
  const section = payload.dueDiligenceCertificatesConsentsAndSignoffs;
  const hasData =
    section.dueDiligenceAreas.length > 0 ||
    section.certificates.length > 0 ||
    section.consents.length > 0 ||
    section.chapterSignoffs.length > 0;
  if (!hasData) return 'not_started';

  const certificatesComplete = section.certificates.every(
    (certificate) => filled(certificate.certificateType) || filled(certificate.provider),
  );
  return certificatesComplete ? 'complete' : 'in_progress';
}

export function evaluateInfrastructureStatus(payload: IntermediariesFilingPayload): SectionStatus {
  const section = payload.depositoriesBankingAsbaUpiAndIssueInfrastructure;
  const core = [
    filled(section.depositoryReadiness.isinStatus),
    filled(section.sponsorBankUpiReadiness.sponsorBankAppointed),
    filled(section.asbaConfiguration.asbaApplicable),
    section.issueBankRoles.length > 0,
  ];
  const answered = core.filter(Boolean).length;
  if (answered === 0) return 'not_started';
  return statusFrom(answered, core.length);
}

export function evaluateUnderwritingStatus(payload: IntermediariesFilingPayload): SectionStatus {
  const section = payload.underwritingMarketMakingAndDistributionArrangements;
  const summary = section.underwritingSummary;
  const hasData =
    filled(summary.issueShares) ||
    filled(summary.totalUnderwritingCommitment) ||
    section.underwritingCommitments.length > 0 ||
    filled(section.marketMakerConfiguration.marketMakerIntermediaryId);
  if (!hasData) return 'not_started';

  const commitmentsComplete = section.underwritingCommitments.every(
    (commitment) => filled(commitment.intermediaryId) || filled(commitment.sharesUnderwritten),
  );
  return commitmentsComplete ? 'complete' : 'in_progress';
}

export function evaluateIssueProgrammeStatus(payload: IntermediariesFilingPayload): SectionStatus {
  const stage = getCurrentFilingStage(payload);
  const section = payload.issueProgrammeAllotmentListingAndPostIssueExecution;

  if (!isStageAtLeast(stage, 'issue_open')) {
    const calendarStarted =
      filled(section.issueCalendar.issueOpeningDate) ||
      filled(section.issueCalendar.issueClosingDate);
    return calendarStarted ? 'in_progress' : 'not_started';
  }

  if (!isStageAtLeast(stage, 'issue_closed')) {
    const openingReadiness = section.issueOpeningReadiness;
    const answered = [
      openingReadiness.rhpProspectusRocFilingReady,
      openingReadiness.pricingFinalized,
      openingReadiness.registrarReady,
    ].filter((value) => value !== '').length;
    return answered === 0 ? 'not_started' : answered < 3 ? 'in_progress' : 'complete';
  }

  if (!isStageAtLeast(stage, 'allotment')) {
    return section.subscriptionRows.length > 0 ? 'in_progress' : 'not_yet_due';
  }

  if (!isStageAtLeast(stage, 'listing_application')) {
    const basis = section.basisOfAllotment;
    const answered = [basis.basisPrepared, basis.allotmentFinalized, basis.exchangeApprovalReceived]
      .filter((value) => value !== '').length;
    return answered === 0 ? 'not_yet_due' : answered < 3 ? 'in_progress' : 'complete';
  }

  const listingStarted =
    filled(section.listing.finalListingApplicationSubmitted) ||
    filled(section.listing.listingDate);
  return listingStarted ? 'in_progress' : 'not_yet_due';
}

export function evaluateFinalDocumentStatus(payload: IntermediariesFilingPayload): SectionStatus {
  const section =
    payload.finalOfferDocumentAdvertisementsMaterialDocumentsAndFilingReadiness;
  const confirmations = section.finalConfirmations;
  const answered = IF_CONFIRMATION_FIELDS.filter((field) => confirmations[field.key] !== '').length;
  if (answered === 0 && section.offerDocumentVersions.length === 0) return 'not_started';
  if (answered < IF_CONFIRMATION_FIELDS.length) return 'in_progress';
  return 'complete';
}

const SECTION_EVALUATORS: Record<
  IntermediariesFilingSectionId,
  (payload: IntermediariesFilingPayload) => SectionStatus
> = {
  'issue-team-and-intermediary-master': evaluateIssueTeamStatus,
  'issue-configuration-and-filing-snapshot': evaluateIssueConfigurationStatus,
  'filing-and-regulatory-milestone-tracker': evaluateFilingTrackerStatus,
  'due-diligence-certificates-consents-and-signoffs': evaluateDueDiligenceStatus,
  'depositories-banking-asba-upi-and-issue-infrastructure': evaluateInfrastructureStatus,
  'underwriting-market-making-and-distribution-arrangements': evaluateUnderwritingStatus,
  'issue-programme-allotment-listing-and-post-issue-execution': evaluateIssueProgrammeStatus,
  'final-offer-document-advertisements-material-documents-and-filing-readiness':
    evaluateFinalDocumentStatus,
};

function deriveOverallStatus(sections: Record<IntermediariesFilingSectionId, SectionStatus>): SectionStatus {
  const values = Object.values(sections);
  const actionable = values.filter(
    (status) => status !== 'not_yet_due' && status !== 'not_applicable',
  );
  if (actionable.length === 0) return 'not_started';
  const completeCount = actionable.filter((status) => status === 'complete').length;
  if (completeCount === 0) return 'not_started';
  if (completeCount === actionable.length) return 'complete';
  return 'in_progress';
}

export function calculateIntermediariesFilingProgress(
  payload: IntermediariesFilingPayload,
): IfProgress {
  const sections = Object.fromEntries(
    (Object.keys(SECTION_EVALUATORS) as IntermediariesFilingSectionId[]).map((sectionId) => [
      sectionId,
      SECTION_EVALUATORS[sectionId](payload),
    ]),
  ) as IfProgress['sections'];

  const sectionsComplete = Object.values(sections).filter((status) => status === 'complete').length;

  return {
    sections,
    sectionsComplete,
    totalSections: Object.keys(sections).length,
    overallStatus: deriveOverallStatus(sections),
    currentFilingStage: getCurrentFilingStage(payload),
  };
}
